src/main/java/org/commcare/suite/model/MenuLoader.java

```diff
@@ -129,39 +131,22 @@ private boolean menuIsRelevant(SessionWrapperInterface sessionWrapper, Menu m) t
         if (m.getMenuRelevance() != null) {
             xPathErrorMessage = m.getMenuRelevanceRaw();
 
-            Set<String> instancesNeededByRelevancyCondition =
-                    (new InstanceNameAccumulatingAnalyzer()).accumulate(relevance);
-            EvaluationContext ec = sessionWrapper.getRestrictedEvaluationContext(m.getId(),
-                    instancesNeededByRelevancyCondition);
-
-            EvaluationContext traceableContext = new EvaluationContext(ec, ec.getOriginalContext());
-            if (traceReporter != null) {
-                traceableContext.setDebugModeOn(traceReporter);
-            }
+            EvaluationContext traceableContext = accumulateInstances(sessionWrapper, m, relevance)
+                    .get("traceableContext");
 
             boolean result = FunctionUtils.toBoolean(relevance.eval(traceableContext));
-            InstrumentationUtils.printAndClearTraces(traceReporter, "menu load expand");
             return result;
         }
         return true;
     }
 
-    public boolean menuAssertionsPass(SessionWrapperInterface sessionWrapper, Menu m) throws XPathSyntaxException{
-        Vector<String> assertionXpathStrings = m.getAssertions().getAssertionsXPaths();
-        if (!assertionXpathStrings.isEmpty()) {
-            XPathExpression assertionXpath = XPathParseTool.parseXPath(assertionXpathStrings.get(0));
+    public boolean menuAssertionsPass(SessionWrapperInterface sessionWrapper, Menu m) throws XPathSyntaxException {
```

Comment:
The code organisation in this method doesn't make much sense as of now. We are looping over individual Xpaths and then calling `m.getAssertions().getAssertionFailure(ec);` inside the loop which goes over all assertions again. Also the `ec` is specific to the individual assertion Xpath that you are iterating against and can't be used to evaluate all assertions. 

I think we need to eval individual assertions instead here, so instead of doing `m.getAssertions().getAssertionFailure(ec);` , we wanna do something like - 
````
Text text  = m.getAssertions().evalAssertionAtIndex(i, ec)
 ````